packages/react/menu/src/Menu.tsx

```diff
@@ -621,6 +636,206 @@ const MenuItemIndicator = React.forwardRef((props, forwardedRef) => {
+
+type MenuSubMenuOwnProps = {
+  open?: boolean;
+  onOpenChange?(open: boolean): void;
+  defaultOpen?: boolean;
+};
+
+const MenuSubMenu: React.FC<MenuSubMenuOwnProps> = (props) => {
+  const { children, open: openProp, defaultOpen, onOpenChange } = props;
+  const triggerRef = React.useRef<HTMLDivElement>(null);
+  const [focusFirstItem, setFocusFirstItem] = React.useState(false);
+  const [renderChildren, setRenderChildren] = React.useState(false);
+  const [open = false, setOpen] = useControllableState({
+    prop: openProp,
+    defaultProp: defaultOpen,
+    onChange: onOpenChange,
+  });
+
+  /**
+   * We defer rendering of child elements until ready, this ensures the following:
+   *
+   * - That measurements are applied correctly to children relative to parent
+   * - That dismissable layers are appended in the correct order when using controlled props
+   */
+  useLayoutEffect(() => {
+    setRenderChildren(true);
+  }, []);
+
+  return (
+    <SubMenuProvider
+      contentId={useId()}
+      triggerRef={triggerRef}
+      focusFirstItem={focusFirstItem}
+      onMouseOpen={React.useCallback(() => {
+        setOpen(true);
+        setFocusFirstItem(false);
+      }, [setOpen])}
+      onKeyboardOpen={React.useCallback(() => {
+        setOpen(true);
+        setFocusFirstItem(true);
+      }, [setOpen])}
+    >
+      <Menu open={open} onOpenChange={setOpen}>
+        {renderChildren && children}
+      </Menu>
+    </SubMenuProvider>
+  );
+};
+
+MenuSubMenu.displayName = SUB_MENU_NAME;
+
+/* -------------------------------------------------------------------------------------------------
+ * MenuSubMenuContent
+ * -----------------------------------------------------------------------------------------------*/
+
+const SUB_MENU_CONTENT_NAME = 'MenuSubMenuContent';
+
+type MenuSubMenuContentOwnProps = Omit<
+  Polymorphic.OwnProps<typeof MenuContent>,
+  'portalled' | 'disableOutsidePointerEvents' | 'disableOutsideScroll'
+>;
+type MenuSubMenuContentPrimitive = Polymorphic.ForwardRefComponent<
+  Polymorphic.IntrinsicElement<typeof MenuContent>,
+  MenuSubMenuContentOwnProps
+>;
+
+const MenuSubMenuContent = React.forwardRef((props, forwardedRef) => {
+  const { ...contentProps } = props;
+  const context = useMenuContext(SUB_MENU_CONTENT_NAME);
+  const subMenuContext = useSubMenuContext(SUB_MENU_CONTENT_NAME);
+  const trigger = subMenuContext.triggerRef.current;
+
+  return (
+    <MenuContent
+      side="right"
+      align="start"
+      {...contentProps}
+      portalled
+      ref={forwardedRef}
+      onOpenAutoFocus={composeEventHandlers(contentProps.onOpenAutoFocus, (event) => {
+        const content = event.target as HTMLElement;
+        if (content && subMenuContext.focusFirstItem) {
+          const items = Array.from(content.querySelectorAll(ENABLED_ITEM_SELECTOR));
+          (items[0] as HTMLElement | undefined)?.focus();
+        }
```

Comment:
I don't think this is going to work correctly with hidden items. Was planning to adjust after https://github.com/radix-ui/primitives/pull/618